File: 0049_group_anagrams/sorted_key/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <cassert>

class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) {
        std::unordered_map<std::string, std::vector<std::string>> groups;
        
        for (const std::string& s : strs) {
            std::string key = s;
            std::sort(key.begin(), key.end());
            groups[key].push_back(s);
        }
        
        std::vector<std::vector<std::string>> result;
        for (auto& pair : groups) {
            result.push_back(std::move(pair.second));
        }
        return result;
    }
};

void test() {
    Solution sol;
    std::vector<std::string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    auto result = sol.groupAnagrams(strs);
    assert(result.size() == 3);
    std::cout << "All tests passed!" << std::endl;
}

int main() {
    test();
    return 0;
}

File: 0049_group_anagrams/sorted_key/kotlin/solution.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val groups = linkedMapOf<String, MutableList<String>>()

        for (str in strs) {
            val chars = str.toCharArray()
            chars.sort()
            val key = String(chars)
            groups.getOrPut(key) { mutableListOf() }.add(str)
        }

        return groups.values.toList()
    }
}
